File: 알고리즘/금과 은 운반하기/Wolfram_금과 은 운반하기.swift

import Foundation

func solution(_ a:Int, _ b:Int, _ g:[Int], _ s:[Int], _ w:[Int], _ t:[Int]) -> Int64 {
    var start = 0
    var end = Int(1e15)
    while start < end {
        let mid : Int = (start + end) / 2
        
        var sum = 0
        var gold = 0
        var silver = 0
        
        for i in 0..<g.count {
            if mid >= t[i] {
                let rm = mid - t[i]
                let time : Int = ((rm / 2) / t[i]) + 1
                sum += min(time * w[i], g[i] + s[i])
                gold += min(time * w[i], g[i])
                silver += min(time * w[i], s[i])
            }
        }
        
        if sum >= a + b, gold >= a, silver >= b{
            end = mid
        }else{
            start = mid + 1
        }
    }
    
    return Int64(start)
}

File: 알고리즘/금과 은 운반하기/Wolfram_금과 은 운반하기.c
// 21.09.15 Solved By Wolfram Hwang
// 파라 메트릭 서치문제

#include <stdio.h>

#define MIN(a, b) a < b ? a : b
typedef long long ll;

long long solution(int a, int b, int g[], size_t g_len, int s[], size_t s_len, int w[], size_t w_len, int t[], size_t t_len) {
    ll start = 0;
    ll end = 1e16;
    int i = 0;
    while (start < end) {
        ll mid = (start + end) / 2;

        int totalSum = 0;
        int Gold = 0;
        int Silver = 0;

        for (i = 0; i < g_len; i++){
            if(mid >= t[i]) {
                ll Remain = mid - t[i];//마지막 운전 할만큼 뺴주기
                ll time = ((Remain / 2) / t[i]) + 1;// 운전 횟수
                totalSum += MIN(time * w[i], g[i] + s[i]);
                Gold += MIN(time * w[i], g[i]);
                Silver += MIN(time * w[i], s[i]);
            }
        }
        if (totalSum >= (a + b) && Gold >= a && Silver >= b) {
            end = mid;
        }else {
            start = mid + 1;
        }
    }
    return start;
}
